From 9d3bad1fc90704a02399d6c56fec9001f29e64d2 Mon Sep 17 00:00:00 2001
From: Sai Surya Teja <sssteja@habana.ai>
Date: Tue, 21 Dec 2021 09:33:43 +0200
Subject: [PATCH] [SW-67660] gaudi2_arc: load preboot from offset

With the new flash layout (below), preboot images are located at
8th and 10th sectors to accommodate 128KB preboot binaries.
But primary and secondary image offsets in image pointers can't be
changed as ARC boot rom checks for ppboot at primary/secondary offset +
64KB.

This patch modifies ppboot_flash_image_offset to fetch from the new
preboot binary location in flash.

New layout:
Sector  image
0	EROM_BIN
1	SECURITY_BIN
2	Unused
3	PPBOOT_BIN_1
4	Unused
5	PPBOOT_BIN_2
6	Image pointers ARM
7	UENV (reservd for any internal testing procedure)
8-9	PREBOOT_BIN_1 (128KB)
10-11	PREBOOT_BIN_2 (128KB)

Change-Id: Ifea7182ab05f252735a508613f600a129d9cb823
---

diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index e91149a..a5e5c5c 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -228,9 +228,9 @@
 	}
 
 	if (image_ptr == ARCBOOT_SECONDARY)
-		return spi_flash_readl(FLASH_SECONDARY_OFFSET);
+		return spi_flash_readl(FLASH_PREBOOT_SECONDARY_OFFSET);
 	else
-		return spi_flash_readl(FLASH_PRIMARY_OFFSET);
+		return spi_flash_readl(FLASH_PREBOOT_PRIMARY_OFFSET);
 }
 
 static int ppboot_bringup_preboot(u8 image_ptr)
diff --git a/arch/gaudi2/arc/common.h b/arch/gaudi2/arc/common.h
index be93150..efadc6a 100644
--- a/arch/gaudi2/arc/common.h
+++ b/arch/gaudi2/arc/common.h
@@ -12,6 +12,13 @@
 #define PPBOOT_SIZE			0x4000		/* 16KB */
 #define PREBOOT_SIZE			0x1FC00		/* 127KB */
 
+#define FLASH_EFUSE_OFFSET		(FLASH_SECONDARY_OFFSET + \
+						sizeof(uint32_t))
+#define FLASH_PREBOOT_PRIMARY_OFFSET	(FLASH_EFUSE_OFFSET + \
+						sizeof(uint32_t))
+#define FLASH_PREBOOT_SECONDARY_OFFSET	(FLASH_PREBOOT_PRIMARY_OFFSET + \
+						sizeof(uint32_t))
+
 #define HL_FW_BOOT_PTRS_BASE		0x80000
 #define HL_SPI_LEGACY_PREBOOT_OFF	0xC0000
 
